Add cop to prevent usage of undefined variables

teckwan opened the pull request:
[**[BUG] Éviter `foo = foo`**](https://trello.com/c/EG9v60LZ)

Addition of a new cop to check for the definition of a local variable every time `foo = ... foo ...` occurs. This ensures that the referred variable is defined and the developer is aware of what he/she is doing. (cf. [slack](https://klaxit.slack.com/archives/GBCMJV8CQ/p1637341309083200))

teckwan commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(9.6) [BUG] Éviter `foo = foo`](https://trello.com/c/EG9v60LZ/2679-96-bug-%C3%A9viter-foo-foo)

BuonOmo reviewed (commented):
I feel like there could be another way around that would not need to check for existence of any existing method name. We'd be solely checking if a variable is assigned to something containing its own name, and not defined earlier.

However, this might be a tall task. I think we could discuss that further IRL 🙂 

Oh also, aren't we missing some tests ? For instance, the ones you gave as example !

teckwan commented:
> I feel like there could be another way around that would not need to check for existence of any existing method name. We'd be solely checking if a variable is assigned to something containing its own name, and not defined earlier.
> 
> However, this might be a tall task. I think we could discuss that further IRL 🙂

It's the "not defined earlier" which im having a hard time doing, which is why i resorted to checking on each time a method is defined.

teckwan commented:
> Oh also, aren't we missing some tests ? For instance, the ones you gave as example !

I might be wrong but there doesn't seem to be any tests for cops.... @BuonOmo

teckwan commented:
Tests to be added.. one case is detected in our codebase, https://github.com/klaxit/klaxit-toolbelt/blob/master/commands/klaxit-ruby-scripts/klaxit_api_reset_stripe_data.rb#L22

BuonOmo commented on rubocop-klaxit/cops/no_undefined_variable_usage.rb:
```diff
@@ -0,0 +1,68 @@
+
+      def_node_search :lvar_with_sym, "(lvar %1)"
+
+      # :nodoc:
+      def on_lvasgn(node)
```
```suggestion
      # Single var assignement.
      #  @example `foo = 1`
      #   (lvasgn :foo
      #     (int 1))
      def on_lvasgn(node)
```

BuonOmo commented on rubocop-klaxit/cops/no_undefined_variable_usage.rb:
```diff
@@ -0,0 +1,68 @@
+        add_offense(node)
+      end
+
+      # :nodoc:
+      def on_masgn(node)
```
```suggestion
      # Multiple vars assignment
      # @example `foo, bar = 1, 2`
      #   (masgn
      #     (mlhs
      #       (lvasgn :foo)
      #       (lvasgn :bar))
      #     (array
      #       (int 1)
      #       (int 2)))

      def on_masgn(node)
```

Also, since `masgn` contains `lvasgn`, is there really a purpose of having that method? I feel like it is a bit unnecessary

BuonOmo commented on rubocop-klaxit/cops/no_undefined_variable_usage.rb:
```diff
@@ -0,0 +1,68 @@
+      def defined_before?(current, sym)
+        parent = current.parent
+
+        return false if %i(def defs).include?(current.type) || parent.nil?
```
```suggestion
	# Local variable scope cannot leak outside a method definition. 
        return false if %i(def defs).include?(current.type)
        # Neither it can leak outside of a whole file.
        return false if parent.nil?
```

BuonOmo commented on rubocop-klaxit/cops/no_undefined_variable_usage.rb:
```diff
@@ -0,0 +1,68 @@
+      def on_lvasgn(node)
+        sym = node.children[0]
+
+        return if lvar_with_sym(node, sym).all? { defined_before?(_1, sym) }
```
Here `_1` should be replace since we're not too familiar with `def_node_search`

BuonOmo commented on rubocop-klaxit/cops/no_undefined_variable_usage.rb:
```diff
@@ -0,0 +1,68 @@
+
+module RuboCop
+  module Cop
+    class NoUndefinedVariableUsage < Cop
```
We're definitetly missing a large comment there indicating what we are looking for!

BuonOmo reviewed (commented):
The nice thing to note is that this case is ok (I wrote it and had that it mind at the time)!

I think the PR is missing a lot of comments to make sure one will be able to maintain and improve that easily. I've added some suggestions, however you can add more, on some blocks you struggled on.

teckwan commented:
> I think the PR is missing a lot of comments to make sure one will be able to maintain and improve that easily. I've added some suggestions, however you can add more, on some blocks you struggled on.

Got it!  Added a few comments following how rubocop documents their cops, hopefully things are clearer.

> Also, since `masgn` contains `lvasgn`, is there really a purpose of having that method? I feel like it is a bit unnecessary

Effectively unnecessary, I thought it would be clearer at first due to the different structure of the nodes, but a branch in on_lvasgn is more than sufficient.

I have also added tests, largely inspired from the rubocop repo. @BuonOmo

BuonOmo commented on rubocop-klaxit/spec/spec_helper.rb:
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+require "rubocop"
+require "rubocop/cop/internal_affairs"
```
oh boy

BuonOmo commented on rubocop-klaxit/spec/cops/no_undefined_variable_usage_spec.rb:
```diff
@@ -0,0 +1,110 @@
+  end
+  it "registers an offense for problematic assignments in parent contexts" do
+    expect_offense(<<~RUBY)
+      foo = foo
+      ^^^^^^^^^ Variable used before being defined
+      begin
+        foo = foo
+        ^^^^^^^^^ Variable used before being defined
+        begin
+          foo = foo
+          ^^^^^^^^^ Variable used before being defined
```
Isn't it weird that the offense is registered n times?

teckwan replied:
It is, but since we are using `on_lvasgn`, we'd have to have some sort of mechanism which remembers if it has already been registered, or only signal the outermost context each time (this would require changing the logic a fair bit). I think it's fine leaving it as it is for now, and if this cop is used extensively as cases like this are supposed to be quite rare.. (hopefully)

BuonOmo replied:
@teckwan just thinking a bit about it: we could however remove that particular test, since having a regression here would be a benefit!

teckwan replied:
signaling only outermost context was easier than expected, and I fixed an error. @BuonOmo

BuonOmo replied:
Well done :clap:

I'll let @Quiwin be the second reviewer

BuonOmo reviewed (approved):
One small thing, yet that may be hard to handle, in which case, lets consider that as unnecessary :)

LGTM, thanks for the clear work on both documentation and tests!

Quiwin reviewed (approved):
LGTM! Thanks for the comments
